### PR TITLE
Removed trait implementation boilerplate with macros

### PR DESCRIPTION
**Pull request by WaDelma**
NOTE: This breaks tests by removing `quickcheck_macros`, because it doesn't work new enough nightly for the macro to work.
NOTE: The macro works only with nightly which is unfortunate.

PR for #1

### CODE REVIEW

**Comment by sebcrozet** [comment]
Oh, sorry, I did not push my version of `alga` yet and I changed almost all trait and file names. For example, all basic structures now start with `Abstract` in order to differentiate them from those that start with `Multiplicative` and `Additive` (which are "trait aliases" for the abstract one plus additional bounds to allow operator overloading). 

It is now on the [nexgen](https://github.com/sebcrozet/alga/tree/nexgen) branch if you which to take a look/contribute. But I did not update the examples yet, the documentation is not done as well, and things might still change quite a bit as a continue the implementations on `nalgebra`. Remarks/questions/comments on the ongoing design are welcome by the way.

I'll merge this PR later manually and adapt to the new trait names if you don't which to dig into the new version.

File: src/macros.rs [hunk 1]
```diff
@@ -13,11 +13,97 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-macro_rules! impl_marker {
-    ($M:ty; $($T:ty),* $(,)*) => {
-        $(impl $M for $T { })+
-    }
-}
+#[macro_export]
+macro_rules! impl_marker(
```
Comment (sebcrozet) [inline src/macros.rs]: Would it be possible to add comments telling what each part of the macro does ? It is very hard to understand anything just by reading the code.

**Comment by WaDelma** [comment]
Welp. Thats quite bit of change in your branch :D

I made this pr mostly to just demonstrate the macro and get feedback (It only works on nightly and breaks tests so it's not really ready for merge).

File: src/macros.rs (hunk 1 above)
Comment (WaDelma) [inline src/macros.rs]: I'll try to do that.

**Comment by WaDelma** [comment]
Added some comments. Do they make sense?

**Comment by sebcrozet** [comment]
Yes It is a bit clearer. I think I understand how the macro work now. Basically, you identify what is before and after `where` and put them into two accumulators which are then used by the very first `@para_rec`. (Also cases without where claude or without parameters are handled.)

Though I don't get the names of the macro variables. I guess `$M` stands for "marker" but I don't see what any the others could mean.

**Comment by WaDelma** [comment]
Good point on variable names: The M comes from marker which it was before I changed it. Will rename the variables too.

**Comment by sebcrozet** [comment]
Great! It is significantly easier to follow now.

**Comment by WaDelma** [comment]
I rebased this to your branch.

I tried to make Travis build the example even without nightly... But dev-dependencies are used even for examples.

**Comment by WaDelma** [comment]
Also there seems to be no AbstractVectorSpace... Is this intentional?

**Comment by sebcrozet** [comment]
The `Abstract` traits don't make any assumption regarding operator overloading. While this is reasonable to have magmas, monoids, and pretty much anything in general algebra, I think it is not reasonable for linear algebra.
I don't expect anybody to write generic code for linear algebra without any operator overloading (for ergonomic and clarity reasons).

**Comment by sebcrozet** [comment]
Regarding travis, I think we could get tests and examples working on Stable by using the stable-compliant quickcheck macro (i.e. use the `quickcheck!( ... )` macro instead of `#[quickcheck]`).

**Comment by WaDelma** [comment]
Okay... It just seemed like it was a oversight, because everything else has Abstract* version.

**Comment by sebcrozet** [comment]
Thanks! I'll fix the tests with the `quickcheck!` macros.

**Pull request merged**